### ISSUE TITLE
bert提取

### ISSUE THREAD

**Issue by windrainmoon** [windrainmoon/lllygogogo#13]
import torch
from transformers import BertForTokenClassification, BertTokenizer, AdamW

# 设定设备
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

# 加载预训练的BERT模型和分词器
model_name = 'bert-base-uncased'
model = BertForTokenClassification.from_pretrained(model_name).to(device)
tokenizer = BertTokenizer.from_pretrained(model_name)

# 准备训练数据（此处简化为示例，实际中需要使用标注数据）
# 假设我们有两个标签：'B-slot', 'I-slot'，以及'O'（非词槽）
tags = ['B-slot', 'I-slot', 'O']
tag2id = {tag: id for id, tag in enumerate(tags)}
id2tag = {id: tag for tag, id in tag2id.items()}

# 示例句子及其词槽标注
sentences = ["I want to book a flight to New York"]
slot_labels = [["O", "O", "O", "O", "O", "O", "O", "B-slot", "I-slot", "I-slot", "O"]]

# 准备数据集
def prepare_dataset(sentences, slot_labels):
    encodings = tokenizer(sentences, truncation=True, padding=True, is_split_into_words=True)
    labels = []
    for i, label in enumerate(slot_labels):
        word_ids = encodings.word_ids(batch_index=i)  # Map tokens to their respective word.
        previous_word_idx = None
        label_ids = []
        for word_idx in word_ids:
            if word_idx is None:
                label_ids.append(-100)
            elif word_idx != previous_word_idx:
                label_ids.append(tag2id[label[word_idx]])
            else:
                label_ids.append(tag2id[label[word_idx]])
            previous_word_idx = word_idx
        labels.append(label_ids)
    encodings['labels'] = labels
    return encodings

# 将数据集转换为PyTorch的Dataset
dataset = prepare_dataset(sentences, slot_labels)

# 将数据集转换为DataLoader
dataloader = torch.utils.data.DataLoader(dataset, batch_size=1, shuffle=True)

# 训练模型
optimizer = AdamW(model.parameters(), lr=5e-5)
model.train()
for epoch in range(3):  # 训练3轮
    for batch in dataloader:
        input_ids = batch['input_ids'].to(device)
        attention_mask = batch['attention_mask'].to(device)
        labels = batch['labels'].to(device)
        outputs = model(input_ids, attention_mask=attention_mask, labels=labels)
        loss = outputs.loss
        loss.backward()
        optimizer.step()
        optimizer.zero_grad()

# 保存模型
model.save_pretrained('slot_model')
tokenizer.save_pretrained('slot_model')

# 加载模型用于预测
model = BertForTokenClassification.from_pretrained('slot_model').to(device)
tokenizer = BertTokenizer.from_pretrained('slot_model')

# 预测
def predict(sentence):
    inputs = tokenizer(sentence, return_tensors="pt", truncation=True, padding=True).to(device)
    with torch.no_grad():
        outputs = model(**inputs)
    predictions = torch.argmax(outputs.logits, dim=-1)
    token_labels = []
    for token, prediction in zip(tokenizer.convert_ids_to_tokens(inputs['input_ids'][0]), predictions[0]):
        if token.startswith('##'):
            token_labels[-1] += token[2:]
        else:
            token_labels.append(id2tag[prediction.item()])
    return token_labels

# 测试预测
test_sentence = "I want to book a flight to Los Angeles"
print(predict(test_sentence))

**Comment by windrainmoon** [comment]
from transformers import BertConfig, BertTokenizer, BertForTokenClassification
from transformers import Trainer, TrainingArguments
import torch

# 设置BERT模型配置
config = BertConfig.from_pretrained('bert-base-chinese')
tokenizer = BertTokenizer.from_pretrained('bert-base-chinese')
model = BertForTokenClassification.from_pretrained('bert-base-chinese')

# 定义训练数据和意图/槽位标签
# 这里需要根据实际数据集进行调整
train_data = {
    'text': [],
    'intent': [],
    'slots': []
}

# 定义验证数据和意图/槽位标签
val_data = {
    'text': [],
    'intent': [],
    'slots': []
}

# 数据预处理函数（根据实际数据集格式进行调整）
def preprocess_function(examples):
    # 将文本拆分为令牌
    encoded_input = tokenizer(examples['text'], add_special_tokens=True, return_tensors='pt')
    # 添加意图和槽位标签
    encoded_input['intent'] = torch.tensor([examples['intent']])
    encoded_input['slots'] = tokenizer(examples['slots'], add_special_tokens=True, return_tensors='pt')
    return encoded_input

# 训练参数
training_arguments = TrainingArguments(
    output_dir='./results',          # 输出目录
    num_train_epochs=3,              # 训练轮数
    per_device_train_batch_size=8, # 训练批次大小
    warmup_比例=0.1,                 # 预热比例
    weight_decay=0.01,               # L2正则化系数
    logging_dir='./logs',           # 日志目录
    learning_rate=5e-5               # 学习率
)

# 训练模型
trainer = Trainer(
    model=model,
    args=training_arguments,
    train_dataset=train_data,
    eval_dataset=val_data,
    compute_metrics=compute_metrics_fn
)

# 训练模型
trainer.train()

# 保存最佳模型
trainer.save_model()

# 计算指标的函数（根据实际需要定义）
def compute_metrics_fn(pred):
    # 这里需要实现计算准确率、召回率和F1分数等指标的逻辑
    pass

# 预测函数（根据实际需要定义）
def predict_function(texts):
    # 使用训练好的模型进行预测
    inputs = tokenizer(texts, return_tensors='pt')
    with torch.no_grad():
        outputs = model(**inputs)
    # 提取预测结果
    pred_intents = outputs.logits[:, 0].argmax(dim=-1)
    pred_slots = tokenizer.batch_decode(outputs.slot_logits[0].argmax(dim=-1), skip_special_tokens=True)
    return pred_intents, pred_slots

# 使用模型进行预测
predicted_intents, predicted_slots = predict_function(['你的服务是什么？'])
print(f'Predicted Intent: {" ".join(tokenizer.convert_ids_to_tokens(predicted_intents[0]))}')
print(f'Predicted Slots: {" ".join(predicted_slots[0])}')
